dataplane-webserver/src/backups/s3.rs

```diff
@@ -0,0 +1,296 @@
+use aws_sdk_s3::{
+    presigning::{PresignedRequest, PresigningConfig},
+    Client as S3Client,
+};
+use aws_smithy_types::byte_stream::ByteStream;
+use serde_json::json;
+
+/// Retrieves and parses backup metadata from an S3 object.
+///
+/// This function attempts to fetch a JSON metadata file from S3 and parse its contents.
+/// It handles various S3 error cases and provides appropriate error messages.
+///
+/// # Arguments
+/// * `client` - AWS S3 client for making requests
+/// * `bucket` - Name of the S3 bucket containing the metadata
+/// * `key` - Full path to the metadata file within the bucket
+///
+/// # Returns
+/// * `Ok(Value)` - Parsed JSON metadata as a serde_json Value
+/// * `Err(Error)` - If the metadata:
+///   - Does not exist in S3
+///   - Cannot be read from S3
+///   - Is not valid JSON
+///
+/// # Errors
+/// Returns an `ErrorInternalServerError` with a descriptive message for:
+/// * Missing metadata file (404)
+/// * S3 access errors
+/// * Invalid JSON format
+pub async fn get_backup_metadata(
+    client: &S3Client,
+    bucket: &str,
+    key: &str,
+) -> Result<serde_json::Value, Error> {
+    let resp = client
+        .get_object()
+        .bucket(bucket)
+        .key(key)
+        .send()
+        .await
+        .map_err(|e| {
+            if let Some(e) = e.as_service_error() {
+                if e.is_no_such_key() {
+                    return ErrorInternalServerError(format!("Backup metadata not found: {}", key));
+                }
+            }
+            ErrorInternalServerError(format!("Failed to get backup metadata from S3: {}", e))
+        })?;
+
+    let body = resp.body.collect().await.map_err(|e| {
+        ErrorInternalServerError(format!("Failed to read backup metadata from S3: {}", e))
+    })?;
+
+    serde_json::from_slice(&body.into_bytes())
+        .map_err(|e| ErrorInternalServerError(format!("Failed to parse backup metadata: {}", e)))
+}
+
+/// Analyzes backup metadata and determines the current backup status.
+///
+/// This function examines the status field in the metadata JSON and constructs
+/// an appropriate BackupStatus enum variant. For completed backups, it also
+/// generates a pre-signed download URL.
+///
+/// # Arguments
+/// * `metadata` - JSON metadata from S3 containing backup status information
+/// * `job_id` - Unique identifier for the backup job
+/// * `client` - AWS S3 client for generating pre-signed URLs
+/// * `config` - Application configuration containing timeout settings
+///
+/// # Returns
+/// * `Ok(BackupStatus)` - Current status of the backup:
+///   - `Completed` with download URL and expiry for successful backups
+///   - `Processing` for in-progress backups
+///   - `Failed` with error message for failed backups
+///   - `Unknown` for unrecognized status values
+/// * `Err(Error)` - If required metadata fields are missing or URL generation fails
+///
+/// # Example Metadata Format
+/// ```json
+/// {
+///     "status": "completed",
```

Comment:
Presumably there'd also be a download URL property.

Reply:
I can see how that can be confusing.  That is example of what the `metadata` input would look like.  I will add something that shows examples of the output of `BackupStatus`